### PR TITLE
KEYCLOAK-12881 KEYCLOAK-13099 Update FederatedIdentities and Groups on POST

### PR DESCRIPTION
**Pull request by slaskawi**
[KEYCLOAK-12881](https://issues.redhat.com/browse/KEYCLOAK-12881)
[KEYCLOAK-13099](https://issues.redhat.com/browse/KEYCLOAK-13099)

This Pull Request introduces adding Federated Identities and Groups when sending User Representation through the REST interface.

### CODE REVIEW

**Comment by stianst** [comment]
See my comment on https://issues.redhat.com/browse/KEYCLOAK-13099. Using RepresentationToModel#createUser in UsersResource#createUser as long as it doesn't break anything, would resolve both this issue and KEYCLOAK-13099.

**Review by stianst — COMMENTED** [review]
See https://issues.redhat.com/browse/KEYCLOAK-13099?focusedCommentId=13982735&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13982735

**Comment by slaskawi** [comment]
@stianst I modified the code per your request. Now both Groups and Federated Identities should be supported.

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java [hunk 1]
```diff
@@ -1785,6 +1780,15 @@ public static UserModel createUser(KeycloakSession session, RealmModel newRealm,
         return user;
     }
 
+    public static void createFederatedIdentities(UserRepresentation userRep, KeycloakSession session, RealmModel realm, UserModel user) {
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java]: Can you revert the change to extract this into a separate method? We don't need a public method like this when it's only used within a single place in the same class

**Review by stianst — CHANGES_REQUESTED** [review]

**Comment by stianst** [comment]
LGTM, but would like the refactor to RepresentationToModel removed as it doesn't serve any purpose.

**Comment by slaskawi** [comment]
@stianst It seems using `RepresentationToModel` approach broke a [bunch of tests](https://travis-ci.com/keycloak/keycloak/builds/150814531). 

I investigated a few of them and it seems the main problem is with Required Actions. Previously, when creating a new user, `#updateUserFromRep` method collected all Required Actions and added it to the newly created user. Of course, `RepresentationToModel#createUser` doesn't do it. 

Now the tricky question is how to address this issue? I can modify `RepresentationToModel#createUser`, so that it pulls Required Actions, just like `#updateUserFromRep` did. Another approach is revert this to the previous implementation I proposed in the first attempt.

**Comment by stianst** [comment]
I don't understand what you mean about "collected all Required Actions and added them". Take a look at:
https://github.com/keycloak/keycloak/blob/master/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java#L1736

Looks to be that it is adding required-actions for users.

**Comment by slaskawi** [comment]
@stianst I meant https://github.com/keycloak/keycloak/blob/dfb67c3aa419674230bb8e345c3583e0282138d8/services/src/main/java/org/keycloak/services/resources/admin/UserResource.java#L221-L223

**Comment by slaskawi** [comment]
@stianst I updated this Pull Request and unified the code in `RepresentationToModel#createUser`.

**Comment by slaskawi** [comment]
@stianst I decided to close this Pull Request as unifying `RepresentationToModel#createUser` and `UserResource#updateUserFromRep` is just too much of a headache. The code only looks similar but there are many, very complicated usecases there. Here's a list of arguments, why this needs to be rewritten from scratch (when implementing REST API v2):

- `RepresentationToModel` assumes importing entries. So it takes them as is - with no modifications.
- `updateUserFromRep` assumes you're creating or updating a User, so it applies hell amount of defaults. This method also manipulates user's email and its username a lot. 
- Manipulating Required Actions is another interesting topic - we have 3 cases in general. Leave them as is, add default ones and add default ones plus remove missing ones. That's just too complex.
- Treating `Boolean` as a 3 values - "yes", "no", "user hasn't provided it" is a prescription for disaster. I've been hunting for a test, which disabled my user for a few hours (!!!).
- We need to rebuild `UserTest`. None of the test methods performs a cleanup and adding a new test method is very problematic. If you change the order of invocation of the tests, they start to fail clashing on duplicated Client IDs or querying users from different tests. Implementing anything there is like walking on a mine field. 

I'll create another PR with a simplified implementation.

**Pull request closed without merging**